### PR TITLE
Move Vfolders association from unit to store

### PR DESCRIPTION
**Pull request by phlax**
atm vfolder are associated per unit, but are only ever used at a store level.

this PR will shift the association from unit to store.

if we need a greater granularity in the future this will not preclude that - but it will make vfolders far more performant just now.

### CODE REVIEW

File: pootle/apps/virtualfolder/receivers.py [hunk 1]
```diff
@@ -31,125 +30,74 @@ def update_vfolder_tree(vf, store):
...
+        if store_added:
+            update_vfolder_tree(vf, store)
 
 
-@receiver(pre_save, sender=VirtualFolder)
-def vfolder_unit_priority_presave_handler(**kwargs):
+@receiver(post_save, sender=VirtualFolder)
+def vfolder_save_handler(sender, instance, created, **kwargs):
     """Remove Units from VirtualFolder when vfolder changes
-
```
Comment (unho) [inline pootle/apps/virtualfolder/receivers.py]: Nit: Keep this blank line.

File: pootle/apps/virtualfolder/migrations/0009_set_vfolder_stores.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+from __future__ import unicode_literals
+
+from django.db import migrations, models
+
+
+def set_vfolder_stores(app, schema):
+    VirtualFolder = app.get_model("virtualfolder.VirtualFolder")
+    Store = app.get_model("pootle_store.Store")
+    for vf in VirtualFolder.objects.all():
+        store_pks = vf.units.values_list("store", flat=True).distinct()
```
Comment (unho) [inline pootle/apps/virtualfolder/migrations/0009_set_vfolder_stores.py]: Just wondering if `store_id` will prevent from an unnecessary extra JOIN.

Comment (phlax) [inline pootle/apps/virtualfolder/migrations/0009_set_vfolder_stores.py]: ill debug_sql

File: pootle/apps/virtualfolder/receivers.py [hunk 3]
```diff
@@ -31,125 +30,74 @@ def update_vfolder_tree(vf, store):
...
     instance = kwargs["instance"]
-    stores = Store.objects.filter(
-        id__in=instance.units.values_list("store_id").distinct())
-    for store in stores:
+    for store in instance.stores.all():
         store.set_priority()
 
 
-@receiver(pre_save, sender=Unit)
-def vfolder_unit_resurrected(**kwargs):
```
Comment (unho) [inline pootle/apps/virtualfolder/receivers.py]: We probably will want to add some test to ensure a vfolder stats are changed when a unit is resurrected or obsoleted.

File: pootle/apps/virtualfolder/migrations/0009_set_vfolder_stores.py (hunk 2 above)
Comment (phlax) [inline pootle/apps/virtualfolder/migrations/0009_set_vfolder_stores.py]: with the following setup:
```
        import pdb; pdb.set_trace()
        from pootle.core.debug import debug_sql
        with debug_sql():
            store_pks = vf.units.values_list("store", flat=True).distinct()
            stores = Store.objects.filter(pk__in=store_pks)
            vf.stores.add(*stores)
```


i get this

```
(Pdb) c
2016-09-10 12:13:18,765 DEBUG 0.001
2016-09-10 12:13:18,765 DEBUG 	QUERY = u'SELECT "pootle_store_store"."id", "pootle_store_store"."file", "pootle_store_store"."pootle_path", "pootle_store_store"."name", "pootle_store_store"."file_mtime", "pootle_store_store"."state", "pootle_store_store"."creation_time", "pootle_store_store"."last_sync_revision", "pootle_store_store"."obsolete", "pootle_store_store"."parent_id", "pootle_store_store"."translation_project_id", "pootle_store_store"."filetype_id", "pootle_store_store"."is_template", "pootle_store_store"."priority" FROM "pootle_store_store" WHERE "pootle_store_store"."id" IN (SELECT DISTINCT U0."store_id" FROM "pootle_store_unit" U0 INNER JOIN "virtualfolder_virtualfolder_units" U1 ON ( U0."id" = U1."unit_id" ) WHERE U1."virtualfolder_id" = %s) ORDER BY "pootle_store_store"."pootle_path" ASC' - PARAMS = (1,)
2016-09-10 12:13:18,765 DEBUG 0.000
2016-09-10 12:13:18,765 DEBUG 	QUERY = u'SELECT "virtualfolder_virtualfolder_stores"."store_id" FROM "virtualfolder_virtualfolder_stores" WHERE ("virtualfolder_virtualfolder_stores"."virtualfolder_id" = %s AND "virtualfolder_virtualfolder_stores"."store_id" IN (%s, %s, %s, %s, %s, %s, %s))' - PARAMS = (1, 1, 2, 3, 15, 16, 17, 30)
```

Comment (phlax) [inline pootle/apps/virtualfolder/migrations/0009_set_vfolder_stores.py]: which im thinking means we want the table join anyway in order to get the stores

Comment (phlax) [inline pootle/apps/virtualfolder/migrations/0009_set_vfolder_stores.py]: i think we dont see any update actually - because there is now signal handlers so the migration didnt add anything

File: pootle/apps/virtualfolder/receivers.py (hunk 3 above)
Comment (phlax) [inline pootle/apps/virtualfolder/receivers.py]: yep - i havent really gotten to looking at stats in this PR yet

File: pootle/apps/virtualfolder/receivers.py [hunk 4]
```diff
@@ -31,125 +30,74 @@ def update_vfolder_tree(vf, store):
...
-        removed_units = (
-            Unit.objects.filter(store__pootle_path__startswith=location,
-                                vfolders=original))
-        for unit in removed_units.iterator():
-            unit.vfolders.remove(original)
-            stores_affected.add(unit.store)
-
-    for store in stores_affected:
-        update_vfolder_tree(original, store)
+            locations.add("".join([location, filename]))
```
Comment (unho) [inline pootle/apps/virtualfolder/receivers.py]: It is probably good idea to move this piece of code that joins together file filter and expanded locations together to the VirtualFolder model. That way all code that deals with location and filtering is kept in the model so it is easier to refactor/expand.

File: tests/models/virtualfolder.py [hunk 5]
```diff
@@ -152,52 +151,25 @@ def test_vfolder_with_no_filter_rules():
...
-    vfolder.location = "/language0/{PROJ}/"
-    vfolder.save()
-
-    expected_units = live_units.filter(
-        store__translation_project__language__code="language0",
-        store__name="store0.po")
-
-    # check vfolder membership after changing the location
+def test_vfolder_membership(tp0, store0):
+    tp0_stores = ",".join(
```
Comment (unho) [inline tests/models/virtualfolder.py]: I would rename it to `tp0_stores_relative_paths`. Otherwise it is misleading.

File: tests/models/virtualfolder.py [hunk 6]
```diff
@@ -152,52 +151,25 @@ def test_vfolder_with_no_filter_rules():
...
+        p[len(tp0.pootle_path):]
```
Comment (unho) [inline tests/models/virtualfolder.py]: Use `to_tp_relative_path`?

File: pootle/apps/virtualfolder/receivers.py (hunk 4 above)
Comment (phlax) [inline pootle/apps/virtualfolder/receivers.py]: its going to the pathmatcher in the next PR

File: pootle/apps/virtualfolder/receivers.py [hunk 7]
```diff
@@ -6,14 +6,12 @@
...
-from django.db.models.signals import post_save, pre_save
+from django.db.models.signals import post_save
 from django.dispatch import receiver
 
-from pootle_store.constants import OBSOLETE
-from pootle_store.models import Store, Unit
+from pootle_store.models import Store
 
 from .models import VirtualFolder, VirtualFolderTreeItem
-from .signals import vfolder_post_save
```
Comment (unho) [inline pootle/apps/virtualfolder/receivers.py]: Is this signal still used?

Comment (phlax) [inline pootle/apps/virtualfolder/receivers.py]: no, ill remove

File: tests/models/virtualfolder.py [hunk 8]
```diff
@@ -152,52 +151,25 @@ def test_vfolder_with_no_filter_rules():
...
-        == sorted(expected_units.values_list("pk", flat=True)))
-
-    vfolder.location = "/language0/{PROJ}/"
-    vfolder.save()
-
-    expected_units = live_units.filter(
-        store__translation_project__language__code="language0",
-        store__name="store0.po")
-
-    # check vfolder membership after changing the location
```
Comment (unho) [inline tests/models/virtualfolder.py]: It would be great to keep this test.

Comment (phlax) [inline tests/models/virtualfolder.py]: k - ill add it back

File: tests/models/virtualfolder.py [hunk 9]
```diff
@@ -314,3 +286,17 @@ def test_vfolder_calc_priority(settings, store0):
...
+    tp0_stores = ",".join(
+        p[len(tp0.pootle_path):]
+        for p in tp0.stores.values_list("pootle_path", flat=True))
+    vf0 = VirtualFolder.objects.create(
+        name="vf0",
+        title="the vf0",
+        location=tp0.pootle_path,
+        filter_rules=tp0_stores)
+    for vfti in vf0.vf_treeitems.iterator():
+        pass
```
Comment (unho) [inline tests/models/virtualfolder.py]: Also test that stats are correctly updated after:

- Changing the units in some store (obsoleting, resurrecting, removing, adding)
- Changing the stores in the vfolder (removing, adding)

Comment (phlax) [inline tests/models/virtualfolder.py]: yep - i was struggling to get how stats are gotten - but ill check again

Comment (unho) [inline tests/models/virtualfolder.py]: We can postpone.

File: pootle/apps/virtualfolder/models.py [hunk 10]
```diff
@@ -131,74 +123,8 @@ def save(self, *args, **kwargs):
...
-                        self.units.add(*qs)
-                        vfolder_stores_set.update(Store.objects.filter(
-                            pootle_path__startswith=vf_file
-                        ))
-
-        # For each store create all VirtualFolderTreeItem tree structure up to
-        # its adjusted vfolder location.
-        for store in vfolder_stores_set:
-            try:
-                VirtualFolderTreeItem.objects.get_or_create(
```
Comment (unho) [inline pootle/apps/virtualfolder/models.py]: This effectively removes the ability to create the VFTI trees for a vfolder. Do we want to include this as part of this PR? Definitely it doesn't belong to this commit.

Comment (unho) [inline pootle/apps/virtualfolder/models.py]: Just discussed this. This is now being handled by signals `VirtualFolder.post_save` and `Store.post_save`.

File: tests/models/virtualfolder.py (hunk 6 above)
Comment (phlax) [inline tests/models/virtualfolder.py]: these are about to go

**Comment by unho** [comment]
lgtm
